#### 1. Last occurence of a char, string

In [27]:
s = 'abcdefge'

def lastOccurence(s: str, ch: chr, i: int, ans: int) -> int:
    # Base Case
    
    if i == len(s):
        return ans

    if s[i] == ch:
        ans = i

    recANs =  lastOccurence(s, ch, i+1, ans)
    return recANs

def lastOccurence_l(s: str, i: int, ch: chr) -> int:
    # base case
    if s[i] == ch:
        return i

    if i == 0:
        return -1

    return lastOccurence_l(s, i-1, ch)

print(lastOccurence(s, 'b', 0, -1))
print(lastOccurence_l(s, len(s)-1, 'z'))

1
-1


#### 2. Reverse a string recursively

In [31]:
s = 'abcdefgz'
s = list(s)
print(s)
def reversing(s: list[str], i: int):
    
    if i == len(s)//2:
        return
    
    s[i], s[-1-i] = s[-1-i], s[i]

    reversing(s, i+1)

reversing(s,0)
print(s)

['a', 'b', 'c', 'd', 'e', 'f', 'g', 'z']
['z', 'g', 'f', 'e', 'd', 'c', 'b', 'a']


#### 3. Add strings Leetcode -> 415

In [75]:
num1 = "999"
num2 = "1"

def solve(num1: str, num2: str, ans: str, carry: str) -> str:

    # Base Case
    if num1 == '' and num2 == '':
        return str(carry) + ans if carry != 0 else ans


    # Processing and recurrence relation
    if num1 == '' or num2 == '':
        if num1 == '':
            ans_digit = carry + int(num2[-1])
            carry = ans_digit // 10
            ans = str(int(ans_digit) % 10)  + ans


            # ans = str(int(num2[-1]) + carry) + ans
            # carry = (int(num2[-1]) + carry) // 10
        else:
            ans_digit = carry + int(num1[-1])
            carry = ans_digit // 10
            ans = str(int(ans_digit) % 10)  + ans


            # ans = str(int(num1[-1]) + carry) + ans
            # carry = (int(num1[-1]) + carry) // 10
        return solve(num1[:-1],num2[:-1],ans,carry)
    else:
        ans_digit = str(carry + int(num1[-1]) + int(num2[-1]))
        carry = int(ans_digit) // 10
        ans = str( int(ans_digit) % 10) + ans
        # print(ans,ans_digit)
    # print(ans,carry)
    recAns = solve(num1[:-1],num2[:-1],ans,carry)
    return recAns


print(solve(num1,num2,'',0))

1000


In [ ]:
num1 = "999"
num2 = "1"

def solve(num1: str, num2: str, ans: str, carry: int) -> str:

    # Base Case
    if num1 == '' and num2 == '':
        return str(carry) + ans if carry != 0 else ans

    # Processing and recurrence relation, i solve one case below, where it touches all exceptions
    ans_digit = str(
        carry + 
        int(num1[-1] if num1 != '' else '0') +
        int(num2[-1] if num2 != '' else '0')               
                    )
    carry = int(ans_digit) // 10
    ans = str( int(ans_digit) % 10) + ans

    # rest is fixed by recAns. I send back ans which solve fill up and then return in recAns, so i return
    # only the recAns, not ans as it's for to pass into solve to build upon the ans
    recAns = solve(num1[:-1] if num1 != '' else '' ,num2[:-1] if num2 != '' else '',ans,carry)
    return recAns


print(solve(num1,num2,'',0))

1000


#### 4. Check Palindrome Recursion

In [ ]:
s = "adca"

def solve(s: str) -> bool:
    
    # Base Case
    if s == '':
        return True
    
    # Processing and recursion relation
    ans = s[0] == s[-1]
    recAns = solve(s[1:len(s)-1])
    ans = ans & recAns

    # Below is also correct but above is more step by step approach, 
    # we solve one case, then got rest of the ans from recursion, and then get total ans from recusrion relation
    # return that ans
    # if we get one false, we return false and ans keep updating as false.

    # if s[0] == s[-1]:
    #     ans = solve(s[1:len(s)-1])
    # else:
    #     ans = False

    return ans


print(solve(s))


False


#### 5. Print all subarrays

In [109]:
nums = [1,2,3,4,5]

def solve(nums: list[int], start: int, end: int):

    if end == len(nums) or start == len(nums):
        return
    
    for i in range(start, end+1):
        print(nums[i],end='')
    print('')

    solve(nums, start, end+1)


    return

for i in range(0,len(nums)):
    solve(nums,i,i)

1
12
123
1234
12345
2
23
234
2345
3
34
345
4
45
5


#### 6. Remove all occurences of the substring. Leetcode 1910


In [1]:
s = 'daabcacadbbcaca'
sub = 'bca'
s = list(s)

l = []
l.append(s[0:2])
print(l)

[['d', 'a']]


In [14]:
s = 'ixcupqoixcupqokevnpokevnpoknqywmlhevgc'
sub = 'ixcupqokevnpo'

def solve(s: list[str], sub: list[str], i: int) -> list[str]:

    # Base Case
    if i == len(s):
        return s

    # Processing
    if s[ i : i + len(sub) ] == sub:
        s = s[0: i] + s[ i + len(sub): ] 
        recAns = (solve(s,sub,max(i - (len(sub) - 1),0)))
    else: 
        recAns = solve(s,sub,i+1)
    
    
    
    return recAns

print("".join(solve(list(s), list(sub), 0)))

knqywmlhevgc


#### 7. Best time to buy and sell stock. Leetcode -> 210


Two pointer solution is actually easy: but I don't understand the story
here basically if there is profit, keep increase sell day,
if profit is negative, move buy day forward
keep sell day ahead of buy day, and update profit if profit is higher than previous one

In [ ]:
#### Two pointer solution is actually easy: but I don't understand the story
#### here basically if there is profit, keep increase sell day,
#### if profit is negative, move buy day forward
#### keep sell day ahead of buy day, and update profit if profit is higher than previous one

def maxProfit(self, prices: List[int]) -> int:
        buy_day = 0
        sell_day = 1
        profit = 0

        if len(prices) == 1:
            return 0

        while(sell_day < len(prices)):
            temp_profit = prices[sell_day] - prices[buy_day]

            if temp_profit > profit:
                profit = temp_profit
                sell_day += 1
            elif temp_profit > 0:
                sell_day += 1
            else:
                buy_day += 1
                if buy_day >= sell_day:
                    sell_day = buy_day + 1

        return profit

In [3]:
#### Two pointer solution is actually easy: but I don't understand the story
#### here basically if there is profit, keep increase sell day,
#### if profit is negative, move buy day forward
#### keep sell day ahead of buy day, and update profit if profit is higher than previous one

prices = [7,1,5,3,6,4]

def maxProfit(prices: list[int], profit, buy_day: int, sell_day: int) -> int:

    if sell_day == len(prices):
        return profit
    

    temp_profit = prices[sell_day] - prices[buy_day]
    if temp_profit > profit:
        profit = temp_profit
        sell_day += 1
    elif temp_profit > 0:
        sell_day += 1
    else:
        buy_day += 1
        if buy_day >= sell_day:
            sell_day = buy_day + 1

    profit = maxProfit(prices, profit, buy_day, sell_day)
    
    return profit

print("Profit:",maxProfit(prices, 0, 0, 1))

Profit: 5


#### 8. 273. Integer to English Words (Leetcode) (REDO)

I couldn't do the solution, got stuck at not being able to resolve 1
1. I found out nearest number.
2. Then tried to get first and second part.
3. But base condditions returns 1, and i cannot resolve it. If i put zero there there is a problem of unlimited recusrsion.

The next code is Lakshay's:
1. Search for nearest number, in the key so the get a build number, then we build the front part of it.
2. we only need to build first part when build no is greater or equal to 100, below that there is no necessity. Kitne 100, 1000 not kitne 80
3. So we get b string from loop, then a string from recursion only when it's above or equal to 100, otherwise it will remain a blank string, 87 -> eighty seven always, nothing in front.
4. Now to get c string (later part) we pass in the rest of the number, only when we need to find it, for example 100 is One Hundred, there is no later part, so we check in num % build_no not zero and only then try to find c string through recursion otherwise c = ''
5. We string ans each time as that is when individual words without other parts are formed, and add a + b + c with spaces, and we return that as an answer.
6. As for the base condition, it's a little hard to figure out.
7. if num == 0 return 0. But this only works in num in first go is zero, otherwise for later part when num == 0, recursion call is never made, instead we return the recursion ans at below, not at base condition. 
8. So Base condition only works in exception case when num = 0 is given in the question. Otherwise and return ans
only works.

As for time complexity -> 
tc -> no. of words
no. of words -> no. of digits ( 2 or less words per digit )
no. of digits -> seil of log<sub>10</sub><sup>num</sup>

for 123 -> 
ciel of (log<sub>10</sub><sup>123</sup>) -> 
ciel of (2.09) -> 2

similar for space complexity.


saving the md format : log<sub>10</sub><sup>num</sup>+log<sub>b</sub><sup>y</sup>

In [30]:
num = 2000000

nword = (
    ('Million', 1000000),
    ('Thousand', 1000),
    ('Hundred', 100),
    ('Ninety', 90),
    ('Eighty', 80),
    ('Seventy', 70),
    ('Sixty', 60),
    ('Fifty', 50),
    ('Forty', 40),
    ('Thirty', 30),
    ('Twenty', 20),
    ('Nineteen', 19),
    ('Eighteen', 18),
    ('Seventeen', 17),
    ('Sixteen', 16),
    ('Fifteen', 15),
    ('Fourteen', 14),
    ('Thirteen', 13),
    ('Twelve', 12),
    ('Eleven', 11),
    ('Ten', 10),
    ('Nine', 9),
    ('Eight', 8),
    ('Seven', 7),
    ('Six', 6),
    ('Five', 5),
    ('Four', 4),
    ('Three', 3),
    ('Two', 2),
    ('One', 1)
)

ndict = {
    
    1000000: 'Million',
    1000: 'Thousand',
    100: 'Hundred',
    90: 'Ninety',
    80: 'Eighty',
    70: 'Seventy',
    60: 'Sixty',
    50: 'Fifty',
    40: 'Forty',
    30: 'Thirty',
    20: 'Twenty',
    19: 'Nineteen',
    18: 'Eighteen',
    17: 'Seventeen',
    16: 'Sixteen',
    15: 'Fifteen',
    14: 'Fourteen',
    13: 'Thirteen',
    12: 'Twelve',
    11: 'Eleven',
    10: 'Ten',
    9: 'Nine',
    8: 'Eight',
    7: 'Seven',
    6: 'Six',
    5: 'Five',
    4: 'Four',
    3: 'Three',
    2: 'Two',
    1: 'One'
}


def isPower(x, y):
    # The only power of 1 is 1 itself
    if x == 1:
        return y == 1

    # Repeatedly compute power of x
    pow = 1
    while pow < y:
        pow *= x

    # Check if power of x becomes y
    return pow == y



def solve(num: int, nword: dict, ans: list[str]) -> list[str ]:

    if num <= 1:
        return ans
    
    build_no = 0
    
    for _, value in nword:
        if value <= num:
            build_no = value
            # print("hi")
            break
    print(build_no)
    # ans.append(build_no)
    
    front_number = num // build_no
    print(front_number,num,build_no)
    # print(front_number)
    tmp = []
    # if front_number > 0:
    front_number_list = solve(front_number, nword, tmp)
    print(front_number_list)
    # print(front_number_list)
    # if front_number_list:
    #     ans.extend(front_number_list)
    # else:
    #     ans.append(1)
    ans.extend(front_number_list)
    # elif front_number == 1:
    #     ans.extend([1])
    
    ans.append(build_no)
    num = num % build_no

    
    
    recAns = solve(num, nword, ans)

    return recAns

substrate = solve(num, nword, [])
# print(substrate)

final_ans = []
print(substrate)
for i in range(0,len(substrate)):
    final_ans.append(ndict[substrate[i]])

print(final_ans)

1000000
2 2000000 1000000
2
1 2 2
[]
[2]
[2, 1000000]
['Two', 'Million']


In [31]:
# Lakshay's soltuion
num = 100

nword = {
    
    1000000000 : 'Billion',
    1000000: 'Million',
    1000: 'Thousand',
    100: 'Hundred',
    90: 'Ninety',
    80: 'Eighty',
    70: 'Seventy',
    60: 'Sixty',
    50: 'Fifty',
    40: 'Forty',
    30: 'Thirty',
    20: 'Twenty',
    19: 'Nineteen',
    18: 'Eighteen',
    17: 'Seventeen',
    16: 'Sixteen',
    15: 'Fifteen',
    14: 'Fourteen',
    13: 'Thirteen',
    12: 'Twelve',
    11: 'Eleven',
    10: 'Ten',
    9: 'Nine',
    8: 'Eight',
    7: 'Seven',
    6: 'Six',
    5: 'Five',
    4: 'Four',
    3: 'Three',
    2: 'Two',
    1: 'One'
}



def solve(num: int, nword: list[tuple]) -> str:
    
    # Base Condition
    if num == 0:
        return 'Zero'

    # Processing and recursive relation.

    # First we find out nearest keyword, from which we will start building number.
    build_no = 0
    for key, _ in nword.items():
        if key <= num:
            build_no = key
            break

    b = nword[build_no]
    # Kitne buld_no, aur kitne hundreds ? kitne thousands ? kitne millions ? but only for num >= 100
    # There is no, ktie 90, kitne 80, only 97 or 83, but in hundred 300 -> kitne hundreds ?:3 hundreds.
    a = ''
    if num >= 100:
        a = solve(num // build_no, nword)

    c = ''
    if num % build_no != 0:
        c = solve(num % build_no, nword)

    ans = a + ' ' +  b + ' ' + c

    return ans.strip(' ')
    
print(solve(num, nword))

One Hundred


#### 9. Wildcard Matching. Leetcode 44

Include / Disclude recursion approach. 

In [35]:
p = ''
if not p: 
    print("yes")

yes


In [69]:
s = 'acdcb'
p = 'a*c?b'




def solve(s: str, p: str) -> bool:

    # Base condition
    if s == p:
        return True
    elif (not s) and ('*' in p):
        return True
    elif (not p) ^ (not s):
        return False

    # Processing and recursive relation
    i = 0
    while ( ( i < len(s)) and (i < len(p) ) ) and (s[i] == p[i] or p[i] == '?'):
        i += 1

    # exclude or after * not consuming anything

    if i < len(p) and p[i] == '*':
        # print(s[i:], p[i+1:], '   ', s[i+1:],p[i:])
        excludeAns = solve(s[i:], p[i+1:])
        # include or after * consuming one character
        includeAns = solve(s[i+1:], p[i:])
        return includeAns or excludeAns
    else:
        return False




print(solve(s,p))

True


#### 10. Number of Dice Roll With Target Sum. Leetcode -> 1155

Lakshays' solution:
We need to see all possible solutions, p&c where n dices with k faces can reach target
1. We will pass target - i th of k faces, when we reach target and exhaust all our dices ie. n == 0, we return 1
and add it to the sum. All other cases we reach till the end, we return 0.
2. We'll run a loop to go through all the faces, and upon calling, pass them in recursion, and return, then add the answer.
3. Basically we choose a phase, and pass that phase to recursion who will fetch it's answer and put in in answer. 

In [1]:
n = 3
k = 6
target = 14

def solve(n: int, k: int, target: int) -> int:

    # Base Case
    if n == 0 and target == 0:
        return 1
    elif ((target != 0) ^ (n != 0)) or target < 0:
        return 0
   


    sum = 0
    # Processing and recurrence relation
    for i in range(1,k+1):
        # print(n-1,k,target-i)
        sum += solve(n-1,k,target-i)
        # print('sum', sum)
        # print(n-1,k,target-i)

    return sum


print(solve(n,k,target))

15
